File: samlang-checker/src/commonTest/kotlin/samlang/checker/TypeResolutionTest.kt
package samlang.checker

import io.kotlintest.shouldBe
import io.kotlintest.shouldThrow
import io.kotlintest.specs.StringSpec
import samlang.ast.common.Type

class TypeResolutionTest : StringSpec() {
    init {
        "can resolve basic disjoint types with sufficient information." {
            val resolution = TypeResolution()
            resolution.addTypeResolution(undecidedTypeIndex = 0, decidedType = Type.int) shouldBe Type.int
            resolution.addTypeResolution(undecidedTypeIndex = 1, decidedType = Type.bool) shouldBe Type.bool
            resolution.addTypeResolution(undecidedTypeIndex = 2, decidedType = Type.string) shouldBe Type.string
            resolution.addTypeResolution(undecidedTypeIndex = 3, decidedType = Type.unit) shouldBe Type.unit
            // Basic single element resolution.
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 0)) shouldBe Type.int
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 1)) shouldBe Type.bool
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 2)) shouldBe Type.string
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 3)) shouldBe Type.unit
            // Recursive resolution.
            resolution.resolveType(
                unresolvedType = Type.FunctionType(
                    argumentTypes = listOf(
                        Type.TupleType(
                            mappings = listOf(
                                Type.UndecidedType(index = 0),
                                Type.UndecidedType(index = 1)
                            )
                        ),
                        Type.UndecidedType(index = 2),
                        Type.FunctionType(
                            argumentTypes = emptyList(),
                            returnType = Type.unit
                        )
                    ),
                    returnType = Type.FunctionType(
                        argumentTypes = emptyList(),
                        returnType = Type.UndecidedType(index = 3)
                    )
                )
            ) shouldBe Type.FunctionType(
                argumentTypes = listOf(
                    Type.TupleType(mappings = listOf(Type.int, Type.bool)),
                    Type.string,
                    Type.FunctionType(argumentTypes = emptyList(), returnType = Type.unit)
                ),
                returnType = Type.FunctionType(argumentTypes = emptyList(), returnType = Type.unit)
            )
        }
        "can link together different type set." {
            val resolution = TypeResolution()
            resolution.addTypeResolution(undecidedTypeIndex = 0, decidedType = Type.int) shouldBe Type.int
            val simpleMeet = { t1: Type, t2: Type ->
                if (t1 == t2) t1 else error(message = "Inconsistency detected")
            }
            resolution.establishAliasing(
                undecidedType1 = Type.UndecidedType(index = 0),
                undecidedType2 = Type.UndecidedType(index = 1),
                meet = simpleMeet
            ) shouldBe Type.int
            resolution.addTypeResolution(undecidedTypeIndex = 2, decidedType = Type.bool) shouldBe Type.bool
            shouldThrow<IllegalStateException> {
                resolution.establishAliasing(
                    undecidedType1 = Type.UndecidedType(index = 0),
                    undecidedType2 = Type.UndecidedType(index = 2),
                    meet = simpleMeet
                )
            }
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 0)) shouldBe Type.int
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 1)) shouldBe Type.int
            resolution.resolveType(unresolvedType = Type.UndecidedType(index = 2)) shouldBe Type.bool
        }
    }
}

File: samlang-core/checker/__tests__/type-resolution.test.ts
import { DummySourceReason } from '../../ast/common-nodes';
import {
  SamlangType,
  SourceBoolType,
  SourceFunctionType,
  SourceIntType,
  SourceStringType,
  SourceUnitType,
} from '../../ast/samlang-nodes';
import { assert } from '../../utils';
import TypeResolution from '../type-resolution';

describe('type-resolution', () => {
  it('can resolve basic disjoint types with sufficient information', () => {
    const resolution = new TypeResolution();
    expect(resolution.addTypeResolution(0, SourceIntType(DummySourceReason))).toEqual(
      SourceIntType(DummySourceReason)
    );
    expect(resolution.addTypeResolution(1, SourceBoolType(DummySourceReason))).toEqual(
      SourceBoolType(DummySourceReason)
    );
    expect(resolution.addTypeResolution(2, SourceStringType(DummySourceReason))).toEqual(
      SourceStringType(DummySourceReason)
    );
    expect(resolution.addTypeResolution(2, SourceStringType(DummySourceReason))).toEqual(
      SourceStringType(DummySourceReason)
    );
    expect(resolution.addTypeResolution(3, SourceUnitType(DummySourceReason))).toEqual(
      SourceUnitType(DummySourceReason)
    );

    // Basic single element resolution.
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 0 })
    ).toEqual(SourceIntType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 1 })
    ).toEqual(SourceBoolType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 2 })
    ).toEqual(SourceStringType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 3 })
    ).toEqual(SourceUnitType(DummySourceReason));
    // Recursive resolution.
    expect(
      resolution.resolveType(
        SourceFunctionType(
          DummySourceReason,
          [
            { type: 'UndecidedType', reason: DummySourceReason, index: 0 },
            { type: 'UndecidedType', reason: DummySourceReason, index: 1 },
            { type: 'UndecidedType', reason: DummySourceReason, index: 2 },
            SourceFunctionType(DummySourceReason, [], SourceUnitType(DummySourceReason)),
          ],
          SourceFunctionType(DummySourceReason, [], {
            type: 'UndecidedType',
            reason: DummySourceReason,
            index: 3,
          })
        )
      )
    ).toEqual(
      SourceFunctionType(
        DummySourceReason,
        [
          SourceIntType(DummySourceReason),
          SourceBoolType(DummySourceReason),
          SourceStringType(DummySourceReason),
          SourceFunctionType(DummySourceReason, [], SourceUnitType(DummySourceReason)),
        ],
        SourceFunctionType(DummySourceReason, [], SourceUnitType(DummySourceReason))
      )
    );
  });

  it('can link together diffSamlangTypet typeSamlangType', () => {
    const resolution = new TypeResolution();
    expect(resolution.addTypeResolution(0, SourceIntType(DummySourceReason))).toEqual(
      SourceIntType(DummySourceReason)
    );

    function simpleMeet(t1: SamlangType, t2: SamlangType) {
      assert(t1 === t2, 'Inconsistency detected');
      return t1;
    }

    expect(
      resolution.establishAliasing(
        { type: 'UndecidedType', reason: DummySourceReason, index: 0 },
        { type: 'UndecidedType', reason: DummySourceReason, index: 1 },
        simpleMeet
      )
    ).toEqual(SourceIntType(DummySourceReason));

    expect(resolution.addTypeResolution(2, SourceBoolType(DummySourceReason))).toEqual(
      SourceBoolType(DummySourceReason)
    );
    expect(() =>
      resolution.establishAliasing(
        { type: 'UndecidedType', reason: DummySourceReason, index: 0 },
        { type: 'UndecidedType', reason: DummySourceReason, index: 2 },
        simpleMeet
      )
    ).toThrow();

    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 0 })
    ).toEqual(SourceIntType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 1 })
    ).toEqual(SourceIntType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 2 })
    ).toEqual(SourceBoolType(DummySourceReason));

    expect(
      resolution.establishAliasing(
        { type: 'UndecidedType', reason: DummySourceReason, index: 3 },
        { type: 'UndecidedType', reason: DummySourceReason, index: 4 },
        simpleMeet
      )
    ).toEqual({ type: 'UndecidedType', reason: DummySourceReason, index: 3 });
    expect(
      resolution.establishAliasing(
        { type: 'UndecidedType', reason: DummySourceReason, index: 2 },
        { type: 'UndecidedType', reason: DummySourceReason, index: 3 },
        simpleMeet
      )
    ).toEqual(SourceBoolType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 2 })
    ).toEqual(SourceBoolType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 3 })
    ).toEqual(SourceBoolType(DummySourceReason));
    expect(
      resolution.resolveType({ type: 'UndecidedType', reason: DummySourceReason, index: 4 })
    ).toEqual(SourceBoolType(DummySourceReason));
  });
});
